opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
 
-                for (int c = 0; c < nc; ++c) {
-                    const int region = fipnum[c] - 1;
-                    if (region != -1 && mask[c]) {
-                        const auto& intQuants = *ebosSimulator_.model().cachedIntensiveQuantities(c, /*timeIdx=*/0);
-                        const auto& fs = intQuants.fluidState();
-                        const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
-                        hcpv[region] += pv[c] * hydrocarbon;
-                        pres[region] += pv[c] * fs.pressure(FluidSystem::oilPhaseIdx).value();
-                    }
+                unsigned cellIdx = elemCtx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
+                const int region = fipnum[cellIdx] - 1;
+                if (region != -1) {
+                    const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
+                    const auto& fs = intQuants.fluidState();
+                    const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
+                    const double pv =
+                        ebosSimulator_.model().dofTotalVolume(cellIdx)
+                        * ebosSimulator_.problem().porosity(cellIdx);
+                    hcpv[region] += pv * hydrocarbon;
+                    pres[region] += pv * fs.pressure(FluidSystem::oilPhaseIdx).value();
                 }
+            }
 
-                comm.sum(hcpv.data(), hcpv.size());
-                comm.sum(pres.data(), pres.size());
+            elemIt = elemCtx.gridView().template begin</*codim=*/0>();
+            for (; elemIt != elemEndIt; ++elemIt) {
+                const auto& elem = *elemIt;
+                if (elem.partitionType() != Dune::InteriorEntity)
+                    continue;
 
-                for (int c = 0; c < nc; ++c) {
-                    const int region = fipnum[c] - 1;
-                    if (region != -1 && mask[c]) {
-                        fip_.fip[FIPData::FIP_PV][c] = pv[c];
-                        const auto& intQuants = *ebosSimulator_.model().cachedIntensiveQuantities(c, /*timeIdx=*/0);
-                        const auto& fs = intQuants.fluidState();
-                        const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
-
-                        if (hcpv[region] != 0) {
-                            fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][c] = pv[c] * fs.pressure(FluidSystem::oilPhaseIdx).value() * hydrocarbon / hcpv[region];
-                        } else {
-                            fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][c] = pres[region] / pv[c];
-                        }
+                elemCtx.updatePrimaryStencil(elem);
+                elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
```

Comment:
the reason is primary that the "new" style works unconditionally, i.e. it works regardless of whether the cached intensive quantities are present or not (or in fact regardless of whether the intensive quantities are cached in the first place)